### PR TITLE
[Improve](map) support map cast with map literal and implicate nested scala cast 

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
support map cast with map literal and implicate nested scala cast 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run buildall

File: regression-test/suites/query_p0/sql_functions/cast_function/test_cast_map_function.groovy [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_cast_map_function", "query") {
+    def tableName = "tbl_test_cast_map_function"
```
Comment (morrySnow) [inline regression-test/suites/query_p0/sql_functions/cast_function/test_cast_map_function.groovy:19]: add some legacy planner cases?

Comment (amorynan) [reply regression-test/suites/query_p0/sql_functions/cast_function/test_cast_map_function.groovy:19]: u mean disable nereids ?

**Comment by amorynan** [comment]
run buildall

File: regression-test/suites/query_p0/sql_functions/cast_function/test_cast_map_function.groovy (hunk 1 above)
Comment (amorynan) [reply regression-test/suites/query_p0/sql_functions/cast_function/test_cast_map_function.groovy:20]: add case more

**Comment by amorynan** [comment]
run feut

**Comment by amorynan** [comment]
run buildall

File: regression-test/suites/nereids_function_p0/cast_function/test_cast_map_function.groovy [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_cast_map_function", "query") {
+    sql """ set enable_nereids_planner = true; """
```
Comment (morrySnow) [inline regression-test/suites/nereids_function_p0/cast_function/test_cast_map_function.groovy:19]: should also set  `enable_fallback_to_original_planner=false`

**Comment by amorynan** [comment]
run buildall

File: regression-test/suites/nereids_function_p0/cast_function/test_cast_map_function.groovy (hunk 2 above)
Comment (amorynan) [reply regression-test/suites/nereids_function_p0/cast_function/test_cast_map_function.groovy:19]: donw!

**Review by morrySnow — APPROVED** [review]

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 3]
```diff
@@ -2169,6 +2169,17 @@ public static boolean matchExactType(Type type1, Type type2, boolean ignorePreci
                     return false;
                 }
                 return matchExactType(((ArrayType) type2).getItemType(), ((ArrayType) type1).getItemType());
+            } else if (type2.isMapType()) {
+                // For types array, we also need to check contains null for case like
```
Comment (xiaokang) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:2173]: map

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 4]
```diff
@@ -2169,6 +2169,17 @@ public static boolean matchExactType(Type type1, Type type2, boolean ignorePreci
                     return false;
                 }
                 return matchExactType(((ArrayType) type2).getItemType(), ((ArrayType) type1).getItemType());
+            } else if (type2.isMapType()) {
```
Comment (xiaokang) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:2172]: Why we need special branch for decimal,array,map? Can we encapsulate all logic in each type's matchesType() or a new method?

File: fe/fe-common/src/main/java/org/apache/doris/catalog/MapType.java [hunk 5]
```diff
@@ -192,8 +192,10 @@ protected String prettyPrint(int lpad) {
     }
 
     public static boolean canCastTo(MapType type, MapType targetType) {
-        return Type.canCastTo(type.getKeyType(), targetType.getKeyType())
-            && Type.canCastTo(type.getValueType(), targetType.getValueType());
+        return (targetType.getKeyType().isStringType() && type.getKeyType().isStringType()
```
Comment (xiaokang) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/MapType.java:195]: 1. add parenthesis for more clear AND OR logic
2. Is already true for StringType canCastTo any type?

**Review by xiaokang — COMMENTED** [review]
add explanation for matchExactType call chain in fe code.

File: fe/fe-common/src/main/java/org/apache/doris/catalog/MapType.java (hunk 5 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/MapType.java:195]: 2. yes , and be will handle this

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**